xfer/model_handler/model_handler.py

```diff
@@ -43,103 +44,484 @@ def __init__(self, module, context_function=mx.context.cpu, num_devices=1, data_
         self.devices = self._get_devices(context_function, num_devices)
         self.data_name = data_name
 
-    def drop_layer_top(self, num_layers_to_drop=1):
+    def drop_layer_top(self, num_layers_to_drop=1, keep_branch_names=None):
         """
         Remove layers from output of model.
 
         :param int n: Number of layers to remove from model output.
+        :param list[str] keep_branch_names: In cases where the top layer is ambiguous because there are branches, this
+                                         should contain the name of the top layer of the branch to keep.
         """
+        sym = self.symbol
+
         self._assert_drop_layer_valid(num_layers_to_drop)
-        # Get index of new last layer
-        last_layer_idx = -(num_layers_to_drop+1)
-        # symbol ending at new last layer
-        sym = self.symbol.get_internals()[self.layer_names[last_layer_idx] + '_output']
-        logging.info('{} deleted from model top'.format(', '.join(self.layer_names[last_layer_idx + 1:])))
+        self._assert_model_has_single_output(self._get_symbol_dict(sym))
+
+        # If keep_branch_names is not a list it can cause unexpected behaviour
+        if keep_branch_names is not None:
+            assert type(keep_branch_names) == list, 'keep_branch_names must be a list of strings'
+
+        layers_dropped = []
+        for n in range(num_layers_to_drop):
+            # Get updated symbol dictionary
+            symbol_dict = self._get_symbol_dict(sym)
+
+            drop_layer_name = symbol_dict[consts.NODES][-1][consts.NAME]  # Get name of last layer
+            last_layer_input_names = self._get_names_of_inputs_to_layer(symbol_dict=symbol_dict,
+                                                                        node_idx=len(symbol_dict[consts.NODES]) - 1)
+            new_last_layer = self._get_layer_name_ambiguous(available_layer_names=last_layer_input_names,
+                                                            reference_layer_names=keep_branch_names, n=n)
+
+            # If case is not ambiguous then new last layer is simply the penultimate layer
+            if new_last_layer is None:
+                next_operation = False
+                for node in reversed(symbol_dict[consts.NODES]):
+                    if next_operation:
+                        if node[consts.OPERATION] != consts.NO_OP:
+                            new_last_layer = node[consts.NAME]
+                            break
+                    if node[consts.NAME] == drop_layer_name:
```

Comment:
By definition, `drop_layer_name` is the last node in `symbol_dict`. This `if` branch to check for `drop_layer_name` and `next_operation` boolean flag can be avoided by directly skipping the last node in `for` statement itself. Use simple code whenever possible.